working/0546-patterns/patterns-feature-specification.md

```diff
@@ -144,953 +144,1073 @@ double calculateArea(Shape shape) =>
-is null, a runtime exception is thrown:
+Dart's existing C-style variable declaration syntax makes it harder to
+incorporate patterns. Variables can be declared just by writing their type, and
+a single declaration might declare multiple variables. Fully incorporating
+patterns into that could lead to confusing syntax like:
 
 ```dart
-(String, Map<String, List<DateTime>?>) data = ...
-var (name, timeStamps!) = data;
+(int, String) (n, s) = (1, "str");
+final (a, b) = (1, 2), c = 3, (d, e);
 ```
 
-### Refutable patterns ("matchers")
-
-Refutable patterns determine if the value in question matches or meets some
-predicate. This answer is used to select appropriate control flow in the
-surrounding construct. Matchers can only appear in a context where control flow
-can naturally handle the pattern failing to match.
+To avoid this weirdness, patterns only occur in variable declarations that begin
+with a `var` or `final` keyword. Also, a variable declaration using a pattern
+can only have a single declaration "section". No comma-separated multiple
+declarations like:
 
-```
-matcher ::=
-  | literalMatcher
-  | constantMatcher
-  | wildcardMatcher
-  | listMatcher
-  | mapMatcher
-  | recordMatcher
-  | variableMatcher
-  | extractMatcher
-  | nullCheckMatcher
+```dart
+var [a] = [1], (b, c) = (2, 3);
 ```
 
-#### Literal matcher
+Also, declarations with patterns must have an initializer. This is not a
+limitation since the point of using a pattern in a variable declaration is to
+match it against the initializer's value.
 
-A literal pattern determines if the value is equivalent to the given literal
-value.
+Add this new rule:
 
 ```
-literalMatcher ::=
-  | booleanLiteral
-  | nullLiteral
-  | numericLiteral
-  | stringLiteral
+patternDeclaration ::= ( 'final' | 'var' ) pattern '=' expression
 ```
 
-Note that list and map literals are not in here. Instead there are list and map
-*patterns*.
+It is a compile-time error if the outermost pattern in a `patternDeclaration`
+is not one of:
 
-**Breaking change**: Using matcher patterns in switch cases means that a list or
-map literal in a switch case is now interpreted as a list or map pattern which
-destructures its elements at runtime. Before, it was simply treated as identity
-comparison.
+*   [`listPattern`][listPattern]
+*   [`mapPattern`][mapPattern]
+*   [`recordPattern`][recordPattern]
+*   [`extractorPattern`][extractorPattern]
+*   [`nullCheckPattern`][nullCheckPattern]
 
-```dart
-const a = 1;
-const b = 2;
-var obj = [1, 2]; // Not const.
+This allows useful code like:
 
-switch (obj) {
-  case [a, b]: print("match"); break;
-  default: print("no match");
-}
+```dart
+var [a, b] = [1, 2];                 // List.
+var {1: a} = {1: 2};                 // Map.
+var (a, b, x: x) = (1, 2, x: 3);     // Record.
+var Point(x: x, y: y) = Point(1, 2); // Extractor.
+var string? = nullableString;        // Null-check.
```

Comment:
This one would not be allowed, because the pattern is refutable. The pattern doesn't match `null`, so `var string? = null;` would not bind `string`, which is not OK for a declaration.

Did you mean `var string! = ...`?

Reply:
Good catch. Wrapped in a pattern-if to make it valid.